src/protocol/fabric/FabricInbox.sol

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {IBlobRefRegistry} from "../../blobs/IBlobRefRegistry.sol";
+import {IProposerFees} from "../IProposerFees.sol";
+import {IPublicationFeed} from "../IPublicationFeed.sol";
+
+import {EIP4788} from "./EIP-4788.sol";
+import {IRegistry} from "urc/src/IRegistry.sol";
+import {BLS} from "urc/src/lib/BLS.sol";
+
+contract FabricInbox {
+    struct Metadata {
+        uint256 anchorBlockId;
+        bytes32 anchorBlockHash;
+        bool isDelayedInclusion;
+    }
+
+    // Contracts
+    IPublicationFeed public immutable publicationFeed;
+    IProposerFees public immutable proposerFees;
+    IRegistry public immutable registry;
+    IBlobRefRegistry public immutable blobRefRegistry;
+
+    // Publication ID trackers
+    uint64 public unsafeHead;
+    uint64 public safeHead;
+
+    // URC-related parameters
+    address public immutable slasher;
+    uint256 public immutable requiredCollateralWei;
+
+    // EIP-4788
+    address private constant beaconRootsContract = 0x000F3df6D732807Ef1319fB7B8bB8522d0Beac02;
+    uint256 private immutable GENESIS_TIMESTAMP;
+
+    // attributes associated with the publication
+    uint256 private constant METADATA = 0;
+    uint256 private constant LAST_PUBLICATION = 1;
+    uint256 private constant BLOB_REFERENCE = 2;
+    uint256 private constant VALIDATOR_PUBKEY = 3;
+
+    constructor(
+        address _publicationFeed,
+        address _blobRefRegistry,
+        address _registry,
+        address _slasher,
+        address _proposerFees,
+        uint256 _requiredCollateralWei,
+        uint256 _genesisTimestamp
+    ) {
+        publicationFeed = IPublicationFeed(_publicationFeed);
+        registry = IRegistry(_registry);
+        slasher = _slasher;
+        proposerFees = IProposerFees(_proposerFees);
+        blobRefRegistry = IBlobRefRegistry(_blobRefRegistry);
+        requiredCollateralWei = _requiredCollateralWei;
+        GENESIS_TIMESTAMP = _genesisTimestamp;
+    }
+
+    // @notice Publish a new publication and update the heads
+    // @param nBlobs The number of blobs being published
+    // @param registrationProof The Merkle proof that msg.sender is registered with the URC
+    // @param validatorProof The EIP-4788 proof that proves who the validator was at a given slot
+    // @param unsafeHeader The PublicationHeader from the previous publication
+    // @param unsafeAttributeHashes The attribute hashes of the previous publication
+    // @param replaceUnsafeHead Whether to replace the unsafe head or update the safe head
```

Comment:
So we only have one branch of unsafe head? I was thinking we can have the unsafe "branches" and retroactively choose one later so the "eventually-safe" branch can't be kicked out by a malicious actor. But I guess that adds the gas cost and complexity?

Reply:
>  so the "eventually-safe" branch can't be kicked out by a malicious actor

I guess this can be mitigated by slashing the malicious actor for submitting an "non-eventually-safe" head. However, without EIP-7917 that introduces the risk of slashing an innocent actor when the lookahead changes due to EB change.